### ISSUE TITLE
subnormal value checker issue

### ISSUE THREAD

**Issue by Tim147** [KhronosGroup/OpenCL-CTS#540]
Hi- 

In the CTS case for vstore_half function, I found a input fp32 value is 6.1005354E-5 (0x387FE000), and it will convert to fp16 format before store. However, this value (6.1005354E-5) is smaller than the smallest fp16 normal format 6.104E-5(0x0400), so it should be subnorml and can be flush to zero. But CTS reference golden is 0x0400, is it reasonable? 

-thanks

**Comment by EwanC** [comment]
Hiya, are you testing on an embedded profile device?

The [spec](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_C.html#built-in-scalar-data-types) under the 'The half Data Type' heading looks like it forbids FTZ behaviour.

> Denormalized numbers for the half data type which may be generated when converting a float to a half using vstore_half and converting a half to a float using vload_half cannot be flushed to zero

However there is an exception in the embedded profile section of the [API spec](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#opencl-embedded-profile), point 7 in section 7.

> Denormalized numbers for the half data type which may be generated when converting a float to a half using variants of the vstore_half function or when converting from a half to a float using variants of the vload_half function can be flushed to zero. The SPIR-V environment specification for details.

**Comment by jlewis-austin** [comment]
Yes, and also it shouldn't be a subnormal after rounding. The original value has 11 significant bits (0x7FE + 1 implicit MSB), but a half subnormal only has 10 (no implicit MSB). So the true value is between the largest half subnormal and the smallest normal. Rounding to discard the extra bit with nearest-even mode will round up to the smallest normal.

**Comment by Tim147** [comment]
> Hiya, are you testing on an embedded profile device?
> 
> The [spec](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_C.html#built-in-scalar-data-types) under the 'The half Data Type' heading looks like it forbids FTZ behaviour.
> 
> > Denormalized numbers for the half data type which may be generated when converting a float to a half using vstore_half and converting a half to a float using vload_half cannot be flushed to zero
> 
> However there is an exception in the embedded profile section of the [API spec](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_API.html#opencl-embedded-profile), point 7 in section 7.
> 
> > Denormalized numbers for the half data type which may be generated when converting a float to a half using variants of the vstore_half function or when converting from a half to a float using variants of the vload_half function can be flushed to zero. The SPIR-V environment specification for details.
Yes, I am testing on embedded device, it can flush to zero for subnormal.

**Comment by Tim147** [comment]
@jlewis-austin thanks, So will the CTS change the subnormal checker to match the spec?

**Comment by jlewis-austin** [comment]
@Tim147 it sounds like that's what's needed here. You're in the best position to propose a PR for it since you'll have to make a patch for your local CTS.

**Comment by rjodinchr** [comment]
Closing as duplicate of #534
